api/models/resourcecleanupstatus.go

```diff
@@ -0,0 +1,98 @@
+package models
+
+import (
+	"fmt"
+	"time"
+)
+
+var resourceCleanupStatusStateTransitions = map[ResourceCleanupStatusState][]ResourceCleanupStatusState{
+	ResourceCleanupStatusStatePending: {
+		ResourceCleanupStatusStateSkipped,
+		ResourceCleanupStatusStateFailed,
+		ResourceCleanupStatusStateDone,
+	},
+}
+
+var resourceCleanupStatusReasonMapping = map[ResourceCleanupStatusState][]ResourceCleanupStatusReason{
+	ResourceCleanupStatusStatePending: {
+		ResourceCleanupStatusReasonAssetScanCreated,
+	},
+	ResourceCleanupStatusStateSkipped: {
+		ResourceCleanupStatusReasonDeletePolicy,
+	},
+	ResourceCleanupStatusStateFailed: {
+		ResourceCleanupStatusReasonProviderError,
+		ResourceCleanupStatusReasonInternalError,
+	},
+	ResourceCleanupStatusStateDone: {
+		ResourceCleanupStatusReasonSuccess,
+	},
+}
+
+func NewResourceCleanupStatus(s ResourceCleanupStatusState, r ResourceCleanupStatusReason, m *string) *ResourceCleanupStatus {
+	return &ResourceCleanupStatus{
+		State:              s,
+		Reason:             r,
+		Message:            m,
+		LastTransitionTime: time.Now(),
+	}
+}
+
+func NewResourceCleanupStatusWithDefaults() *ResourceCleanupStatus {
+	return NewResourceCleanupStatus(ResourceCleanupStatusStatePending, ResourceCleanupStatusReasonAssetScanCreated, nil)
+}
+
+func (rs *ResourceCleanupStatus) Equals(r *ResourceCleanupStatus) bool {
+	if rs.Message == nil && r.Message != nil {
+		return false
+	}
+	if r.Message == nil && rs.Message != nil {
+		return false
+	}
+	if rs.Message == nil && r.Message == nil {
+		return rs.State == r.State && rs.Reason == r.Reason
+	}
+
+	return rs.State == r.State && rs.Reason == r.Reason && *rs.Message == *r.Message
+}
+
+func (rs *ResourceCleanupStatus) isValidStatusTransition(r *ResourceCleanupStatus) error {
+	transitions, ok := resourceCleanupStatusStateTransitions[rs.State]
+	if ok {
+		for _, transition := range transitions {
+			if transition == r.State {
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("invalid transition: from=%s to=%s", rs.State, r.State)
+}
+
+func (rs *ResourceCleanupStatus) isValidReasonTransition(r *ResourceCleanupStatus) error {
```

Comment:
```suggestion
func (rs *ResourceCleanupStatus) isValidReason(r *ResourceCleanupStatus) error {
```

The name for this method is misleading. There is no transition for `reason`. There are valid `reason`(s) for certain `state`(s).